ocaml/xapi/certs.ml

```diff
@@ -0,0 +1,414 @@
+    parse_string ~consume parser str
+
+  let parse_file_using parser path =
+    try read_file path |> parse_string_using parser
+    with e ->
+      Error (Printf.sprintf "Can't process %s: %s" path (Printexc.to_string e))
+end
+
+(* module [Sync] provides an interface to certificates stored on
+disk. Certificates are read upon [init], potentially manipulated, and
+written back on [sync]. No other code should write these files. Writing
+back on [sync] provides a primitive form of transactions: in case of a
+fatal error, we could avoid writing a bad state to disk. This is
+currently not actively used, though *)
+
+module Sync : sig
+  type uuid = Uuidm.t
+
+  type paths = {
+      host_cert_path: string
+    ; ca_certs_dir: string
+    ; ca_crls_dir: string
+  }
+
+  type host = {
+      private_key: X509.Private_key.t
+    ; cert: X509.Certificate.t
+    ; intermediares: X509.Certificate.t list
+  }
+
+  type ca = {cert: X509.Certificate.t; uuid: uuid; mutable deleted: bool}
+
+  type cr = {name: string; cert: X509.Certificate.t; mutable deleted: bool}
+
+  (* [path] value to be used normally but other paths could be used for
+     testing *)
+  val default_paths : unit -> paths
+
+  (* initialise this module and read satte from disk *)
+  val init : paths -> (unit, [> R.msg]) result
+
+  (* write back stack in memory to disk *)
+  val sync : unit -> (unit, [> R.msg]) result
+
+  (* remove CA cert *)
+  val rm_ca : uuid -> unit -> (unit, [> R.msg]) result
+
+  (* remove CR cert *)
+  val rm_cr : string -> unit -> (unit, [> R.msg]) result
+
+  (* access CR cert *)
+  val get_cr : unit -> ((string * X509.Certificate.t) list, [> R.msg]) result
+
+  (* access CA cert *)
+  val get_ca : unit -> ((uuid * X509.Certificate.t) list, [> R.msg]) result
+
+  (* access host cert *)
+  val get_host :
+       unit
+    -> ( (X509.Private_key.t * X509.Certificate.t * X509.Certificate.t list)
+         option
+       , [> R.msg] )
+       result
+end = struct
+  type uuid = Uuidm.t
+
+  type paths = {
+      host_cert_path: string
+    ; ca_certs_dir: string
+    ; ca_crls_dir: string
+  }
+
+  type host = {
+      private_key: X509.Private_key.t
+    ; cert: X509.Certificate.t
+    ; intermediares: X509.Certificate.t list
+  }
+
+  type ca = {cert: X509.Certificate.t; uuid: uuid; mutable deleted: bool}
+
+  type cr = {name: string; cert: X509.Certificate.t; mutable deleted: bool}
+
+  type t = {
+      host: host option
+    ; mutable ca: ca list
+    ; mutable cr: cr list
+    ; paths: paths
+  }
+
+  let default_paths () =
+    {
+      host_cert_path= !Xapi_globs.server_cert_path
+    ; ca_certs_dir= "/etc/stunnel/certs"
+    ; ca_crls_dir= Stunnel.crl_path
+    }
+
+  (* local, private, mutable state *)
+  let t : t option ref = ref None
+
+  (* calls in this API are protected by this mutex *)
+  let mx = Mutex.create ()
+
+  let with_mutex f =
+    Mutex.lock mx ;
+    defer (fun () -> Mutex.unlock mx) f
+
+  let get () =
```

Comment:
Can we call this `get_unsafe`/`get_no_lock`? So that it's obvious it ought to be used inside `with_mutex`, and that we're not going to get a deadlock